File: subprocmgr.py
# subprocmgr - Subprocess manager (Python component)
#
# Copyright © 2015 Zack Weinberg
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# http://www.apache.org/licenses/LICENSE-2.0
# There is NO WARRANTY.

"""This is a semi-compatible replacement for the `subprocess` module.
   Together with its C helper program, it solves a headache when writing a
   Python program that runs a bunch of subprocesses and consumes their
   output: There is no way to select() for the termination of a specific
   process, and therefore no reliable way to know when to call wait().

   The most significant divergence from the `subprocess` API is that
   `Popen` objects do not provide the `poll`, `wait`, or `communicate`
   methods, nor the `stdin`, `stdout`, or `stderr` properties.
   Instead, there is a `status` property, which is a `queue.Queue`;
   data on the queue consists of `SubprocessStatus` objects, which
   carry all of the information that you could have gotten from the
   above methods, but with an ordering guarantee.

   The special constants available for the `stdin`, `stdout`, and
   `stderr` arguments to the `Popen` constructor are different, and
   the default behavior (if these arguments are not specified) is also
   different:

       subprocmgr.DEVNULL - Use /dev/null.
                            This is the default for `stdin`.
       subprocmgr.RECEIVE - Receive output via `status`.
                            This is the default for `stdout` and `stderr`.
       subprocmgr.INHERIT - Inherit this file descriptor.  It will point to
                            whatever it pointed to in the parent process
                            *at the time the C helper program was started*.

   If you want to set a standard stream to a pipe, you have to create
   the pipe manually.  It is strongly recommended that you only do this
   for intermediate channels in a pipeline.

   The `bufsize`, `shell`, `preexec_fn`, `pass_fds`, `close_fds`,
   `restore_signals`, and `start_new_session` arguments to the `Popen`
   constructor are not implemented.  Some of these are implementable
   if needed, others are not.  (The behavior is as-if `close_fds=True`
   and `restore_signals=True`.)

   The new constructor arguments `status_queue` and
   `package_status_message` allow you to supply an existing queue that
   should receive status messages, and package `SubprocessStatus`
   objects if it is necessary to make them fit the queue's format
   (e.g. if it's a priority queue).

   The convenience functions `call`, `check_call`, and `check_output`
   are not implemented.  They wouldn't be hard to add, but if you need
   this module you probably aren't using them anyway.

   The module function `start_manager` can be used to start up the C
   helper program and internal event loop, if this needs to happen
   before the first call to `subprocmgr.Popen` for some reason (e.g.
   if you want to do it before creating threads, or if you are going
   to change what stdin/stdout/stderr point to).  `stop_manager` stops
   them again; this also automatically happens when the program exits.
   Note that invoking `stop_manager` causes all subprocesses that are
   still running to receive a SIGTERM, and five seconds later, any
   survivors receive a SIGKILL; it will not return until all
   outstanding subprocesses have exited.  (Their final output is
   posted to the respective .status queues as usual.)

   Finally, note that the *direct* parent of processes created using this
   module is the C helper program, which itself runs as a subprocess of
   your program.  This could cause problems in edge cases.

   This module currently only works on Linux, because the C helper program
   currently uses Linux-specific APIs (namely, epoll(2) and signalfd(2)).
   Patches to make it work on operating systems which implement kqueue(2)
   will cheerfully be accepted.  Patches for portability to other Unixes
   may or may not be accepted, depending on how horrible the necessary
   contortions are.

   There is probably no hope of making this module work on Windows, but if
   someone is prepared to do all of the work and all of the subsequent
   maintenance, I would at least consider the proposal.  (You will
   *certainly* need to reimplement the helper program from scratch, and
   possibly also the communication channel to this module.)

"""

import atexit
import errno
import fcntl
import logging
import os
import queue
import resource
import socket
import struct
import threading

#
# Public constants
#

# These must not collide with the legitimate fd space (nonnegative
# integers) or with file objects.  We also don't want them to collide
# with the similar constants from the subprocess module.
RECEIVE = "RECEIVE"
INHERIT = "INHERIT"
DEVNULL = "DEVNULL"

#
# Internal utilities
#

_log = logging.getLogger(__name__)
_helper_path = os.path.join(os.path.dirname(__file__), "subprocmgr")

def interpret_wait_status(status):
    """Internal: produce a human-readable message from a wait status."""
    if os.WIFEXITED(status):
        if os.WEXITSTATUS(status) == 0:
            return "exited normally"
        else:
            return "exited with failure code {}".format(os.WEXITSTATUS(status))
    elif os.WIFSIGNALED(status):
        return "killed by signal {}".format(os.WTERMSIG(status))
    else:
        return "posted uninterpretable wait status {:04x}".format(status)

def start_helper_process(control_socket):
    """Internal: start up a helper process connected to 'control_socket',
       and return its PID."""

    pid = os.fork()
    if pid: return pid

    # We are the child.  We want to inherit fds 0, 1, and 2 as is,
    # place the control socket at fd 3, and close all higher-numbered
    # file descriptors.  If this is 3.4, we can rely on the stdlib's
    # general policy of non-inheritance (with the clever exception of
    # the result of a call to dup2()!) to do that for us; otherwise we
    # have to close higher-numbered fds manually, and neither Linux
    # nor Python gives us closefrom(), grumble.
    successful_dup2 = False
    try:
        os.dup2(control_socket.fileno(), 3)
        successful_dup2 = True

        if not hasattr(os, 'set_inheritable'):
            # This is the least terrible way to simulate closefrom()
            # on Linux that I am aware of.  There doesn't seem to be
            # any way to find out the number of the fd used to scan
            # /proc/self/fd (which _will_ be in the list of fds
            # returned) so we just close it too and ignore the error.
            # Creating a list in advance ensures that we won't close
            # that fd out from under os.listdir (which I believe is
            # _not_ a generator in any released version of Python,
            # but that could easily change).
            try:
                fds = [x for x in
                       (int(xx) for xx in os.listdir("/proc/self/fd"))
                       if x >= 4]
            except OSError:
                s, h = resource.getrlimit(resource.RLIMIT_NOFILE)
                fds = range(4, s)

            for fd in fds:
                try: os.close(fd)
                except OSError: pass

        os.execl(_helper_path, "subprocmgr")

        # We should never get here.
        os._exit(127)

    except Exception as e:
        # Write a well-formed status message to the control socket
        # reporting our own failure to start up, and then exit abnormally.
        # Note that if we've already done the dup2() then we should use
        # fd 3 instead of the original fd, in case the original fd has
        # been closed.
        errc = getattr(e, 'errno', -1)
        text = str(e).encode("utf-8")
        msg = struct.pack("=4I{}s".format(len(text)),
                          0,    # pseudo-tag for the helper itself
                                # (code below will never use this)
                          1,    # status: system error during process creation
                          errc, # value: errno
                          len(text),
                          text)
        if not successful_dup2:
            control_socket.sendall(msg)
        else:
            socket.fromfd(3, socket.AF_UNIX, socket.SOCK_STREAM).sendall(msg)
        os._exit(1)

class Manager:
    """Internal: responsible for maintaining the communication channel to
       the C helper program, marshaling messages in both directions, etc.
       Usable as a context manager.
    """

    def __init__(self):
        self._helper_pid = None
        self._helper_skt = None
        self._decode_thr = None
        self._procs = {}
        self._stopping = False

    def __enter__(self):
        self.start_manager()

    def __exit__(self, et, ev, eb):
        self.stop_manager()
        return False

    def start_manager(self):
        if self._helper_pid is not None:
            return
        if self._stopping:
            raise RuntimeError("cannot start manager while it is being stopped")

        # Take care not to leak any resources if initialization fails
        s1 = None
        s2 = None
        pid = None
        thr = None
        try:
            (s1, s2) = socket.socketpair(socket.AF_UNIX, socket.SOCK_STREAM)
            pid = start_helper_process(s2)
            s2.close(); s2 = None
            thr = threading.Thread(
                target=self.decode_and_dispatch_status_messages,
                args=(s1,)
            )
            thr.run()

        except:
            if s1 is not None: s1.close()
            if s2 is not None: s2.close()
            if pid is not None: os.kill(pid, signal.SIGKILL)
            if thr is not None and thr.is_alive(): thr.join()
            raise

        self._helper_pid = pid
        self._helper_skt = skt
        self._decode_thr = thr


    def stop_manager(self):
        if self._helper_pid is None:
            return
        if self._stopping:
            return

        self._stopping = True

        # this notifies the helper program that it should exit
        self._helper_skt.shutdown(socket.SHUT_WR)

        (_, status) = os.waitpid(self._helper_pid, 0)
        if status:
            _log.warning("stop_manager: helper " +
                         interpret_wait_status(status))
        else:
            _log.debug("stop_manager: helper " +
                       interpret_wait_status(status))

        self._decode_thr.join()

        self._decode_thr = None
        self._helper_skt = None
        self._helper_pid = None
        self._stopping = False

    def decode_and_dispatch_status_messages(self, sock):
        """Thread worker procedure: receive status messages from the
           helper program and dispatch them to the appropriate Popen
           object's output queue."""
        ...

    def start_process(self, proc):
        """Actually start up the subprocess described by PROC."""
        ...

#
# Global state: unless otherwise specified, subprocesses created by this
# module are under this manager.  It will be started the first time it is
# needed, or when start_manager() is called.  It can be forcibly stopped
# by calling stop_manager().
#

_default_manager = Manager()
def start_manager():
    """Start the default subprocess manager now."""
    _default_manager.start_manager()

def stop_manager():
    """Stop the default subprocess manager now."""
    _default_manager.stop_manager()

atexit.atexit(stop_manager)

#
# Public classes
#

class SubprocessStatus:
    ...

class Popen:
    ...

File: subprocmgr.c
/* subprocmgr - Subprocess manager (C component).
 *
 * Copyright © 2015 Zack Weinberg
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * There is NO WARRANTY.
 */

/* This program, together with the Python module that uses it, solves
   one very specific headache when writing a Python program that runs
   a bunch of subprocesses and consumes their output: There is no way
   to select() for the termination of a specific process, and
   therefore no reliable way to know when to call wait().

   This program is not meant to be invoked directly.  It communicates
   with the Python module using an internal protocol that is subject
   to change without notice, but which I will document here anyway.
   It takes no arguments, and will never write anything to stdout or
   read anything from stdin.  Human-readable error messages may be
   written to stderr under unusual failure conditions.

   On invocation, file descriptor 3 must be an AF_UNIX/SOCK_STREAM
   socket, referred to as the "control socket", which is used to
   instruct this program to create new subprocesses.  Abstractly,
   there is only one type of message sent to this program on this
   socket, but it's processed as a pair of sub-messages.  The first
   sub-message of each pair is expected to consist of 2 32-bit,
   native-endian integers in this order:

       data_len
       n_fds

   The second sub-message consists of 'data_len' bytes of ordinary
   data, plus 'n_fds' file descriptors as SCM_RIGHTS data.  At
   present, the second sub-message will be ill-formed if it does not
   provide at least 16 bytes of data and one file descriptor.

   The second sub-message's data has the format

       uint32_t   tag
       uint8_t    flags
       uint8_t    disposition of fd 0 in subprocess
       uint8_t    disposition of fd 1 in subprocess
       uint8_t    disposition of fd 2 in subprocess
       uint32_t   argument count
       uint32_t   environment variable count
       cstring    name of executable
       cstring[]  argument vector
       cstring[]  environment vector

   The file descriptors passed, if any, are a simple array, referenced
   by the "disposition of fd N in subprocess" fields of the message.

   'tag' is an arbitrary, invoker-selected value used to distinguish
   processes in status messages.  It is the invoker's responsibility
   not to reuse tags while their associated processes are still alive.

   'flags' is currently reserved and must be all-bits-zero.

   The "disposition of fd N in subprocess" codes have the following
   possible values.  Only the value -1 (== 0xFF) is treated as negative.

       -1         Inherit from parent.
       0          fd 0:   Open /dev/null for read.
                  fd 1/2: Forward output.
       k >= 1     Use passed fd with index (k-1).

   Currently it is not possible to supply higher-numbered file descriptors
   to the child.

   The executable name, argument vector, and environment vector are
   all packed together as a sequence of C-strings; there is no
   formatting other than the NUL-terminators.  The executable name is
   mandatory but the other two are optional.  The argument count and
   environment variable count give the number of entries in their
   respective vectors, with two special cases: If the argument count
   is zero, the executable name will be reused as the sole entry in
   the argument vector passed to execve(2).  If the environment
   variable count is -1 (== 0xFFFF_FFFF), the environment vector is
   expected to be empty, and the new process will inherit its
   environment from the parent (i.e. this program).

   (If the environment variable count is zero, the new process will
   recieve a completely empty environment.)

   This program sends "status messages" back to the invoker via the
   control socket.  These messages consist of 4 unsigned 32-bit
   integers followed by zero or more bytes of data:

       tag
       status
       value
       len

   'tag' is always the tag provided with the message that created the
   process, and 'len' always indicates how many bytes of data follow.
   The meaning of 'value' depends on 'status', which is one of the
   following possible codes:

       0    The control message was ill-formed.  'value' is zero,
            and the data is a human-readable message describing the problem.

       1    System error during process creation. 'value' is an errno code
            and the data is a human-readable error message.  This message
            will include strerror(value).

       2    Process successfully created.  'value' is the process ID.
            No data.

       3    Process has produced output.  'value' will be 1 for stdout
            or 2 for stderr, and the data is a block of output.  This
            program does not reblock or transform the data in any way;
            one chunk of data read from the pipe = one message.

       4    Process has closed an output channel. 'value' will be 1 for
            stdout or '2' for stderr.  No data.

       5    Process has exited.  'value' is the wait status.  No data.

   For any given process, this program guarantees to emit messages in
   the following order: First, exactly one message with status 0, 1,
   or 2.  If the code was 0 or 1, there will be no further messages
   for that tag, and all passed file descriptors have been closed.
   Otherwise, any number of messages with status 3, followed by
   exactly one message with status 4, for whichever of stdout, stderr,
   or both were given disposition "Forward output via the status
   pipe".  (There is no ordering between stdout and stderr.)  Finally,
   exactly one message with status 5.

   When EOF is received on the control socket, this program will send
   SIGTERM to all processes that are still running, and start a
   five-second timer.  It will continue to generate status messages
   until there are no more messages to generate (i.e. all children
   have exited), and then it will exit itself.  If the timer expires,
   any surviving processes receive a SIGKILL and status message
   generation continues.

   If this program ever receives a write error on the control socket,
   all further output from subprocesses, and their wait statuses,
   will be read and discarded.  This condition does *not* cause running
   subprocesses to be terminated.

   If this program ever receives SIGHUP, SIGINT, SIGQUIT, SIGALRM,
   SIGTERM, SIGVTALRM, SIGXCPU, SIGXFSZ, or SIGPWR, it will behave as
   if it had received EOF on the control socket, except that the
   initial signal sent is the same as was received.  If this program
   receives SIGILL, SIGABRT, SIGFPE, SIGBUS, SIGSEGV, SIGSYS, or
   SIGTRAP, it will immediately send SIGKILL to all processes that are
   still running, and then crash as usual for that signal.
   The signals that stop processes are allowed to behave normally.
   All other signals are ignored.

   (Child processes receive whatever signal mask this program's parent
   provided it.)
 */

/* Portability note: this program makes use of many POSIX.1-2008
   (including XSI) APIs and several APIs that are currently
   Linux-specific.  */

#define _GNU_SOURCE
#include <stddef.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <errno.h>

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/epoll.h>
#include <fcntl.h>
#include <unistd.h>
